runtime-sdk/src/crypto/signature/ed25519.rs

```diff
@@ -94,3 +95,76 @@ impl From<PublicKey> for CorePublicKey {
         pk.0
     }
 }
+
+/// A memory-backed signer for Ed25519.
+pub struct MemorySigner {
+    sk: ed25519_dalek::ExpandedSecretKey,
+}
+
+impl MemorySigner {
+    /// Creates a new signer from a RFC 8032 seed.
+    pub fn new_from_seed(seed: &[u8]) -> Result<Self, Error> {
+        let sk = ed25519_dalek::SecretKey::from_bytes(seed).map_err(|_| Error::InvalidArgument)?;
+        let esk = ed25519_dalek::ExpandedSecretKey::from(&sk);
+        Ok(Self { sk: esk })
+    }
+
+    /// Generates a new signer deterministically from a test key name string.
+    pub fn new_test(name: &str) -> Self {
+        let mut digest = Sha512Trunc256::new();
+        digest.update(name.as_bytes());
+        let seed = digest.finalize();
+
+        Self::new_from_seed(&seed).unwrap()
+    }
+
+    /// Reconstruct the signer from a byte representation of the key.
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self, Error> {
+        Ok(Self {
+            sk: ed25519_dalek::ExpandedSecretKey::from_bytes(bytes)
+                .map_err(|_| Error::MalformedPrivateKey)?,
+        })
+    }
+
+    /// Return a byte representation of this signer's key.
+    pub fn to_bytes(&self) -> Vec<u8> {
+        self.sk.to_bytes().to_vec()
+    }
+
+    /// Public key corresponding to the signer.
+    pub fn public(&self) -> PublicKey {
+        let pk = ed25519_dalek::PublicKey::from(&self.sk);
+        PublicKey::from_bytes(pk.as_bytes()).unwrap()
+    }
+
+    /// Generates a signature with the private key over the context and message.
+    pub fn context_sign(&self, context: &[u8], message: &[u8]) -> Result<Signature, Error> {
+        let mut digest = Sha512Trunc256::new();
+        for byte in &[context, message] {
+            digest.update(byte);
+        }
+        let message = digest.finalize();
+
+        let pk = ed25519_dalek::PublicKey::from(&self.sk);
+        let signature = self.sk.sign(&message, &pk);
+
+        Ok(signature.to_bytes().to_vec().into())
```

Comment:
I think returning something other than `Signature` would break the interface a bit too much, since the other implementations also return a Signature. I could extend Signature's interface to make it easier to avoid copying?